File: test-suite-python/src/test/python/micronaut/security/docs/principalparam/UserController.py
# tag::imports[]
from java.security import Principal
from micronaut.http.annotation import Controller, Get
from micronaut.security.annotation import Secured
# end::imports[]
from micronaut.context.annotation import Requires


@Requires(property="spec.name", value="PrincipalParamTest")
# tag::clazz[]

@Controller("/user")
class UserController:

    @Secured("isAnonymous()")
    @Get("/myinfo")
    def myinfo(self, principal: Principal | None) -> dict:
        if principal is None:
            return {"isLoggedIn": False}
        return {"isLoggedIn": True, "username": principal.getName()}
# end::clazz[]

File: test-suite/src/test/java/io/micronaut/security/docs/principalparam/UserController.java
package io.micronaut.security.docs.principalparam;

//tag::imports[]
import io.micronaut.core.util.CollectionUtils;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.security.annotation.Secured;
import org.jspecify.annotations.Nullable;

import java.security.Principal;
import java.util.Collections;
import java.util.Map;
//end::imports[]
import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "PrincipalParamTest")
//tag::clazz[]

@Controller("/user")
public class UserController {

    @Secured("isAnonymous()")
    @Get("/myinfo")
    public Map<String, Object> myinfo(@Nullable Principal principal) {
        if (principal == null) {
            return Collections.singletonMap("isLoggedIn", false);
        }
        return CollectionUtils.mapOf("isLoggedIn", true, "username", principal.getName());
    }
}
//end::clazz[]
